### PR TITLE
Add Azure DevOps dnceng instance CI YAML

### PR DESCRIPTION
**Pull request by dagood**
CentOS, Ubuntu 16.04, and OSX legs included.

Fedora 28 and Debian 8.2 builds don't work yet due to the images we're using missing tools/libs, filing issues to track those. (https://github.com/dotnet/source-build/issues/814, https://github.com/dotnet/source-build/issues/815.)

I haven't tried these new defs with internal changes yet (`user.PAT`). I tried to keep the functionality around so hopefully that will just be a quick check.

https://github.com/dotnet/source-build/issues/757

### CODE REVIEW

**Comment by dagood** [comment]
I temporarily pushed this branch to the `dotnet` source-build repo to get around some agent pool resource authorization issues. (I think since status is commit-based, the build I started based on the `dotnet` branch is showing up here now.)

**Comment by dagood** [comment]
Note the new CI status, [`source-build-CI — #20181015.2 succeeded`](https://dnceng.visualstudio.com/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_build/results?buildId=30849). Unfortunately there's no way to break it out per platform yet for visibility.

**Comment by dagood** [comment]
@dotnet-bot test OSX10.12 Release

File: .vsts.pipelines/phases/ci-linux.yml [hunk 1]
```diff
@@ -1,20 +1,16 @@
...
-- phase: ${{ parameters.name }}
+- phase: ${{ parameters.phase }}
   variables:
-    # Prefix to distinguish artifacts from different legs. No documented variable for this exists.
-    artifactName: no_artifact_name
-    buildLoggingOptions: ''
-    buildConfiguration: Release
-    buildOfflineTarball: false
+    # Prefix to distinguish artifacts from different legs.
+    artifactName: ${{ format('$(type)_{0}', parameters.phase) }}
```
Comment (dseefeld) [inline .vsts.pipelines/phases/ci-linux.yml]: Why an underscore separator here when the type param contains space separators anyway?

File: .vsts.pipelines/phases/ci-osx.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    artifactName: ${{ format('$(type)_{0}', parameters.phase) }}
+    logsDirectory: $(Build.ArtifactStagingDirectory)/logs
+    SOURCE_BUILD_SKIP_SUBMODULE_CHECK: true
+    # Default type, can be overridden by matrix legs.
+    type: Production
+  queue:
+    name: ${{ parameters.queueName }}
+    demands: ${{ parameters.queueDemands }}
+    timeoutInMinutes: 240
+    matrix:
```
Comment (dseefeld) [inline .vsts.pipelines/phases/ci-osx.yml]: I don't understand this syntax.  Why can't this just be: matrix: ${{ parameters.matrix }}  ?

File: .vsts.pipelines/phases/ci-linux.yml (hunk 1 above)
Comment (dagood) [reply .vsts.pipelines/phases/ci-linux.yml]: This is probably leftover from an older understanding of AzDO's naming requirements--changing to space.

File: .vsts.pipelines/phases/ci-osx.yml (hunk 2 above)
Comment (dagood) [reply .vsts.pipelines/phases/ci-osx.yml]: I'll give it a try (I agree it'd make sense if that worked), I think I remember getting a validation error trying to add it like that. The insert syntax is documented here: https://docs.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=vsts#insertion

**Comment by dagood** [comment]
> Is there a plan for updating this to the new Azure DevOps YAML schema in a separate PR?

What new schema do you mean? 😨

File: .vsts.pipelines/phases/ci-osx.yml (hunk 2 above)
Comment (dagood) [reply .vsts.pipelines/phases/ci-osx.yml]: Nevermind, that works fine!

**Comment by dagood** [comment]
Looking closer at the new docs, I'm noticing now that I guess `phase`s are now `job`s, and instead of `phase => matrix` there's `job => strategy => matrix`, maybe more things like that. I'll file an issue for it.

**Pull request merged**